spices/0002-uri-rewriting-for-mirroring.adoc

```diff
@@ -0,0 +1,319 @@
+= SPICE 2: URI Rewriting for Mirroring
+
+* Proposal: link:./0002-uri-rewriting-for-mirroring.adoc[SPICE-0002]
+* Author: link:https://github.com/holzensp[Philip Hölzenspies]
+* Status: Review
+* Implemented in: Pkl 0.26.0
+* Category: Tooling
+
+== Introduction
+
+This SPICE discusses rewrite rules for URIs of modules being `import`ed, so as to use mirrored alternatives.
+
+== Motivation
+
+For systems that require run-time air-gapping, packages imported from `pkg.pkl-lang.org` or other sites on the internet are not available.
+This can be remedied by pre-populating a cache directory with all required dependencies (see link:../spices/0001-import-graph-analyzer.pkl[SPICE 1: Import Graph Analyzer API]).
+When dependencies can not be determined ahead of time, pre-populating the cache is not an option.
+Mirroring packages from internet repositories in a local store is insufficient, because the mirrored packages may have hard-coded dependencies.
+Importing mirrored packages, using URIs pointing to the mirror, will eventually evaluate `import`s of those hard-coded dependencies, referring to external sites.
+
+Another reason to want to consume (transitive) dependencies from a mirror is the requirement of altered versions of packages.
```

Comment:
This amounts to monkey patching. Is it wise for a deterministic config language to encourage/support this?

Reply:
I don't think this is a valid use-case; packages have their integrity verified by checksum. Altering a package would break any integrity checks that we would do.